package/MDAnalysis/lib/distances.py

```diff
@@ -389,6 +389,118 @@ def self_distance_array(reference, box=None, result=None, backend="serial"):
+        The unitcell dimesions for this system format as returned
+        by :attr:`MDAnalysis.coordinates.base.Timestep.dimensions`:
+        ``[lx,ly, lz, alpha, beta, gamma]``. Minimum image convention
+        is applied if the box is provided.
+    method : 'bruteforce' or 'pkdtree' or 'None'
+        Based on benchmarks, checks the suitable method
+        and search the indices of the selections in relatively
+        less amount of time. 'bruteforce'  can be costly with 
+        large number of coordinates. Another option is 
+        periodic KDTree and can be advantageous with increase in particles.
+    
+    Returns
+    -------
+    indices : list
+        Combination of coordinate indices, one from each
+        from reference and configuration state such that
+        distance between them is smaller than cutoff distance
+        provided as ``cutoff`` in arguments. 
+    
+    Note
+    -----
+    Currently only supports brute force. Similar methods 
+    can be defined and can be directly linked 
+    to this function.
+    """
+    
+    if box is not None:
+        boxtype = _box_check(box)
+        # Convert [A,B,C,alpha,beta,gamma] to [[A],[B],[C]]
+        if (boxtype == 'tri_box'):
+            box = triclinic_vectors(box)
+        if (boxtype == 'tri_vecs_bad'):
+            box = triclinic_vectors(triclinic_box(box[0], box[1], box[2]))
+    if method is None:
+        method = _determine_method(reference, configuration, max_cutoff, box = box)
+
+    if method == 'bruteforce':
+        pairs = _bruteforce_capped(reference, configuration, max_cutoff, min_cutoff = min_cutoff, box = box)
+    #if method = 'pkdt':
+    #    pairs = _pkdt_capped(reference, configuration, cutoff, box = None)
+    return pairs    
+    
+
+def _determine_method(reference, configuration, cutoff, box):
+    """
+    Switch between different methods based on the the optimized time.
+    All the rules to select the method based on the input can be 
+    incorporated here.
+
+    Returns
+    -------
+    String of the method. supports ``bruteforce`` or ``pkdt`` as of now
+
+    Note 
+    ----
+    Just a basic rule based on the number of particles
+    """
+    #if (len(reference) > 1000) or (len(configuration) > 1000):
+    #    return 'pkdt'
+    return 'bruteforce'
+
+
+def _bruteforce_capped(reference, configuration, max_cutoff, min_cutoff = 0,  box = None):
+    """
+    Using naive distanc calulations, returns a list 
+    containing the indices with one from each 
+    reference and configuration list, such that the distance between
+    them is less than the specified cutoff distance
+    """
+    pairs = []
+    if reference.shape == (3,):
+        reference = reference[None, :]
+    if configuration.shape == (3,):
+        configuration = configuration[None, :]
+
+    _check_array(reference, 'reference')
+    _check_array(configuration, 'configuration')
+
+    for i, coords in enumerate(reference):
```

Comment:
So this way of iterating over one coordinate set is probably slower than `distance_array(ref, conf)`.  I think `guess_bonds` does it this way to avoid memory problems

Reply:
Yes exactly. The `distance_array` cannot go beyond 5k particles. So unless we are explicitly limiting the brute force algorithm to 5k, shouldn't the more robust way be a part of the calculation. I agree once we put in the rules, the iteration over each particle would be about half as fast as the distance_array and should be replaced.